### PR TITLE
[Api][Address] Browsing AddressBook

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

**Review by AdamKasp — COMMENTED** [review]
generally I have problem with scenario tagged `@javascript`, it looks like UI feature, and IMO we shouldn't cover it with API especially when you added step `And my billing address is fulfilled automatically after filling field which is matching to field from my address book's address` in API we don't have behaviour like `automatically do sth` it should be call to `addresses/index` and the UI  should fill address with data from addressbook, But of course i am open to discussion.

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 1]
```diff
@@ -52,6 +53,9 @@ final class CheckoutContext implements Context
     /** @var ApiClientInterface */
     private $countriesClient;
 
+    /** @var ApiClientInterface */
+    private $addressClient;
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:57]:
```suggestion
    private $addressesClient;
```
to be consistent with other clients in this service

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 2]
```diff
@@ -100,6 +106,21 @@ public function __construct(
         $this->sharedStorage = $sharedStorage;
     }
 
+    /**
+     * @Given /^(my) billing address is fulfilled automatically through default address$/
+     */
+    public function myBillingAddressIsFulfilledAutomaticallyThroughDefaultAddress(UserInterface $user): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:112]:
```suggestion
    public function myBillingAddressIsFulfilledAutomaticallyThroughDefaultAddress(ShopUserInterface $user): void
```
?

File: src/Sylius/Behat/Resources/config/services/api.xml [hunk 3]
```diff
@@ -18,6 +18,11 @@
             <argument type="service" id="sylius.behat.shared_storage" />
         </service>
 
+        <service id="sylius.behat.api_platform_client.shop.address" class="Sylius\Behat\Client\ApiPlatformClient" parent="sylius.behat.api_platform_client">
+            <argument>address</argument>
```
[the remark this answers is not in the document]
Comment (Tomanhez) [reply src/Sylius/Behat/Resources/config/services/api.xml:22]: I'm not sure about this change, because in `UI` we have adressBook with path `addresses`, but generally we have  entity `Address` and in REST API we should create endpoint like `/address`, look in other entities and paths in UI and API

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 4]
```diff
@@ -127,6 +148,17 @@ public function iAmAtTheCheckoutAddressingStep(): void
...
 
+    /**
+     * @When /^I choose "([^"]+)" street for (billing|shipping) address$/
+     */
+    public function iChooseForBillingAddress(string $street, string $addressType): void
+    {
+        $this->sharedStorage->set(
+            'filled_address_field',
+            ['street' => $street, 'addressType' => $addressType . 'Address']
+        );
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:159]: I guess you should here get an address from address book and fill it in content data

File: src/Sylius/Component/Core/Repository/AddressRepositoryInterface.php [hunk 5]
```diff
@@ -19,9 +19,6 @@
 
 interface AddressRepositoryInterface extends RepositoryInterface
 {
-    /**
-     * @return array|AddressInterface[]
-     */
```
Comment (GSadee) [inline src/Sylius/Component/Core/Repository/AddressRepositoryInterface.php:24]: Why did you remove that?

File: src/Sylius/Behat/Resources/config/services/api.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (AdamKasp) [reply src/Sylius/Behat/Resources/config/services/api.xml:22]: check this `bin/console debug:router |grep new-api/admin/` ALL path in our API have resource name in plural form, and all services in `src/Sylius/Behat/Resources/config/services/api.xml` have an argument in the plural form. One of us is wrong and i believe that isn't me :D

**Comment by Tomanhez** [comment]
> generally I have problem with scenario tagged `@javascript`, it looks like UI feature, and IMO we shouldn't cover it with API especially when you added step `And my billing address is fulfilled automatically after filling field which is matching to field from my address book's address` in API we don't have behaviour like `automatically do sth` it should be call to `addresses/index` and the UI should fill address with data from addressbook, But of course i am open to discussion.

I think that few tests for addressBook like `choosing_billing_and_shipping_address_from_address_book.feature` are wrote not perfect for UI, because lots is happening automatically and is based on autocomplete from web browser, I add few new steps for greater clarity and of course for improve to work with @api tests. I know that many of these behawior is hard to recode on api cases but, IMO its only what i can do. Other way is write new tests for API behawior instead of present tests.

File: features/account/customer_account/address_book/viewing_addresses.feature [hunk 6]
```diff
@@ -6,15 +6,15 @@ Feature: Viewing my address book
 
     Background:
         Given the store operates on a single channel in "United States"
-        And I am a logged in customer
+        And I am a logged in customer with name "Lucifer Morningstar"
```
[the remark this answers is not in the document]
Comment (Tomanhez) [reply features/account/customer_account/address_book/viewing_addresses.feature:9]: I changed it, because IMO scenario `Inability to view the addresses of other customers` is not clearly and which is customer about in step `this address should be assigned to "Lucifer Morningstar"`, we have added address with this name for this customer, but this customer is not ` "Lucifer Morningstar"` but random named

File: src/Sylius/Behat/Resources/config/services/api.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tomanhez) [reply src/Sylius/Behat/Resources/config/services/api.xml:22]: Yes you aright, I was looking wrong 🚀

**Comment by GSadee** [comment]
> > generally I have problem with scenario tagged `@javascript`, it looks like UI feature, and IMO we shouldn't cover it with API especially when you added step `And my billing address is fulfilled automatically after filling field which is matching to field from my address book's address` in API we don't have behaviour like `automatically do sth` it should be call to `addresses/index` and the UI should fill address with data from addressbook, But of course i am open to discussion.
> 
> I think that few tests for addressBook like `choosing_billing_and_shipping_address_from_address_book.feature` are wrote not perfect for UI, because lots is happening automatically and is based on autocomplete from web browser, I add few new steps for greater clarity and of course for improve to work with @api tests. I know that many of these behawior is hard to recode on api cases but, IMO its only what i can do. Other way is write new tests for API behawior instead of present tests.

I'm not fully convinced but we could implement the existing scenarios without adding these unneeded (unclear for me) new steps. The proper behaviour for getting address from customer's address book could be implemented in the existing steps in API context.

File: src/Sylius/Behat/Context/Api/Shop/AddressContext.php [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        Assert::same(count($this->responseChecker->getCollection($this->client->getLastResponse())), $count);
+    }
+
+    /**
+     * @Then this address should be assigned to :fullName
+     */
+    public function thisAddressShouldBeAssignedTo(string $fullName): void
+    {
+        /** @var AddressInterface $address */
+        $address = $this->sharedStorage->get('address_assigned_to_' . $fullName);
```
[the remark this answers is not in the document]
Comment (Tomanhez) [reply src/Sylius/Behat/Context/Api/Shop/AddressContext.php:67]: Yes you aright, but `$fullName` is behind the scene, I can convert it to loverCase and add padding here and in places when I must get it by customer name, but for me its unnecessary confusion.

**Review by arti0090 — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Shop/AddressContext.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (arti0090) [reply src/Sylius/Behat/Context/Api/Shop/AddressContext.php:67]: You can use StringInflector class, just add there an method 
````
public static function nameToSnakeCase(string $value): string
    {
        return (string) u($value)->snake();
    }
```

**Comment by AdamKasp** [comment]
Thanks, Tomasz! :tada:

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 8]
```diff
@@ -1023,4 +1090,42 @@ private function removeOrderItemFromCart(int $orderItemId, string $tokenValue):
...
+            $address['street'] === $addressToCompare->getStreet() &&
+            $address['city'] === $addressToCompare->getCity() &&
+            $address['postcode'] === $addressToCompare->getPostcode() &&
+            ($addressToCompare->getProvinceName() !== null && isset($address['provinceName'])) ?
+                $address['provinceName'] === $addressToCompare->getProvinceName() : true
+        ) {
+            return true;
+        }
+
+        return false;
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:1120]: Could be simplified to one return
